File: project_euler/Problem047/Problem047.java
package project_euler;

import java.util.Arrays;

public class Problem047 {
    public void solution() {
        int target = 4;
        int length = 200000;
        int[] factors = new int[length];
        Arrays.fill(factors, -1);

        int count = 0;
        for (int i = 2; i <= length; ++i) {
            if (target == getFactors(i, factors, length)) {
                ++count;
                if (target == count) {
                    System.out.println(i - 3);
                    return;
                }
            } else {
                count = 0;
            }
        }
    }

    private int getFactors(int value, int[] factors, int length) {
        if (-1 == factors[value - 1]) {
            long index = value;
            while (index <= length) {
                factors[(int) index - 1] = 1;
                index *= value;
            }

            for (int i = Math.min(value - 1, length / value); i >= 2; --i) {
                factors[value * i - 1] = factors[value - 1] + factors[i - 1];
            }
        } else {
            for (int i = Math.min(value - 1, length / value); i >= 2; --i) {
                if (isRelativelyPrime(value, i) && -1 == factors[value * i - 1]) {
                    factors[value * i - 1] = factors[value - 1] + factors[i - 1];
                }
            }
        }
        return factors[value - 1];
    }

    private boolean isRelativelyPrime(int i, int j) {
        while (0 != i % j) {
            int auxiliary = i % j;
            i = j;
            j = auxiliary;
        }
        return 1 == j;
    }
}

File: project_euler/Problem047/Problem047.js
function Problem047() {
    let target = 4
    let length = 200000
    let factors = Array(length).fill(-1)

    let count = 0
    for (let i = 2; i <= length; ++i) {
        if (target === getFactors(i, factors, length)) {
            ++count
            if (target === count) {
                console.log(i - 3)
                return
            }
        } else {
            count = 0
        }
    }
}

function getFactors(value, factors, length) {
    if (-1 === factors[value - 1]) {
        let index = value
        while (index <= length) {
            factors[index - 1] = 1
            index *= value
        }

        for (let i = Math.min(value - 1, Math.floor(length / value)); i >= 2; --i) {
            factors[value * i - 1] = factors[value - 1] + factors[i - 1]
        }
    } else {
        for (let i = Math.min(value - 1, Math.floor(length / value)); i >= 2; --i) {
            if (isRelativelyPrime(value, i) && -1 === factors[value * i - 1]) {
                factors[value * i - 1] = factors[value - 1] + factors[i - 1]
            }
        }
    }
    return factors[value - 1]
}

function isRelativelyPrime(i, j) {
    while (i % j) {
        let auxiliary = i % j
        i = j
        j = auxiliary
    }
    return 1 === j
}

module.exports = {
    Problem047 : Problem047
};
